### ISSUE TITLE
Allow inline function expressions in arrow operator call chains

### ISSUE THREAD

**Issue by rhdunn** [qt4cg/qtspecs#436]
It can be useful to create inline functions for simple operations (e.g. adding 1 to a number) to be used in arrow operator call chains.

The current proposal uses `-> { ... }` for just the thin arrow operator.

This proposal is split into two parts:
1. restructure the grammar to make the function call usage simpler to follow;
2. introduce the ability to use inline functions in thin/fat arrow contexts.

## Part 1 -- Simplify the Grammar
I suggest changing the grammar to:
```
FatArrowTarget           ::= "=>" ( ArrowFunctionCall | ArrowDynamicFunctionCall )
ThinArrowTarget          ::= "->" ( ArrowFunctionCall | ArrowDynamicFunctionCall )
ArrowFunctionCall        ::= EQName ArgumentList
ArrowDynamicFunctionCall ::= ( VarRef | ParenthesizedExpr ) PositionalArgumentList
```

That is, I've grouped the function name/reference with the argument list.

## Part 2 -- Allow Inline Functions
```
FatArrowTarget          ::= "=>" ( ArrowFunctionCall | ArrowDynamicFunctionCall | ArrowInlineFunctionCall )
ThinArrowTarget         ::= "->" ( ArrowFunctionCall | ArrowDynamicFunctionCall | ArrowInlineFunctionCall )
ArrowInlineFunctionCall ::= ( "function" | "fun" ) EnclosedExpr
```

Note: here, `"fun"` is a placeholder for whichever name/symbol we choose in https://github.com/qt4cg/qtspecs/issues/53.

This then allows expressions like `(1, 2, 3) -> function { . + 1 }` and `(1, 2, 3) => fun { ~ = 1 }` (see also https://github.com/qt4cg/qtspecs/issues/129) without overloading the meaning of `->`.

**Comment by michaelhkay** [comment]
It feels confusing to me that I sometimes have to write an empty argument list and sometimes not. For example after `let $x := fn { . + 1 }`, I can write

`"fred" -> $x()`

or

`"fred" -> fn { . + 1 }`

but not

`"fred" -> $var`

or

`$fred -> fn { . + 1 }()`

Generally, my feeling is that the design lacks orthogonality.

**Comment by michaelhkay** [comment]
I guess one way to address these concerns might be to say that the "()" on the RHS is not needed if there are no extra arguments to supply, so `"fred" => upper-case()` could be now be written as `"fred" => upper-case`. The ability to omit the argument list then applies uniformly regardless of how the function is supplied.

It would take some fairly careful analysis to be confident this does not introduce any ambiguities or incompatibilities. And even then, is it actually a good thing?

**Comment by ChristianGruen** [comment]
I agree, there’s are good reasons for keeping parentheses whenever you call functions (even though I sympathize with the `"fred" -> fn { . + 1 }` suggestion).

Do we have a chance to get rid of the additional parentheses for inline functions?

```xquery
(: right now :)
$seq -> (function($n) { $n + 1 })()

(: desirable :)
$seq -> function($n) { $n + 1 }()
```

For functions with single arguments, and with the syntax recently in [#435](https://github.com/qt4cg/qtspecs/issues/435#issuecomment-1504810495) and a dot as replacement for `$1`, we could have:

```xquery
$seq -> { . + 1 }()
```

**Comment by ndw** [comment]
The second option isn't desirable to my eyes. Making me work backwards from `()` through `{$n + 1}` to `function($n)` to work out what the `()`'s "bind to" isn't preferable to me. But as I commented a moment ago, I won't be surprised if my desire for explicitness where possible puts me in the minority.

Note also that the paren form is a common idiom from JavaScript:

```javascript
(function() {
  // stuff goes here
})();
```

**Comment by ChristianGruen** [comment]
> Note also that the paren form is a common idiom from JavaScript […]:

Good point.

> The second option isn't desirable to my eyes. Making me work backwards from `()` through `{$n + 1}` to `function($n)` to work out what the `()`'s "bind to" isn't preferable to me. But as I commented a moment ago, I won't be surprised if my desire for explicitness where possible puts me in the minority.

Yes, I see. I could have used `=>` instead of `->` in the examples, but that doesn’t really affect your observation (verbosity contributes to comprehensiveness).

**Comment by ChristianGruen** [comment]
An addendum to:

> Note also that the paren form is a common idiom from JavaScript:
> 
> ```js
> (function() {
>   // stuff goes here
> })();
> ```

I remembered that parentheses are optional in XPath (this is legal: `function($x) { $x }(123)`), but they’re mandatory if expressions are chained, so the main reason here would be to get the syntax aligned (if possible).

**Comment by dnovatchev** [comment]
> so `"fred" => upper-case()` could be now be written as `"fred" => upper-case`

Actually this would need to be:

**`"fred" => upper-case#1`**

Many years ago I asked if it would not be good to omit the arity in any case when a function definition has just a single overload. The answer is no, because if someone in the future adds another overload, then the expression without explicit arity becomes ambiguous

**Comment by rhdunn** [comment]
> I guess one way to address these concerns might be to say that the "()" on the RHS is not needed if there are no extra arguments to supply, so `"fred" => upper-case()` could be now be written as `"fred" => upper-case`. The ability to omit the argument list then applies uniformly regardless of how the function is supplied.

`upper-case` on its own, without function call syntax (`upper-case()`) or a function reference (`upper-case#1`) is hard to read/follow. It looks too much like an element path step.

Q: Is there a reason why named function references are not allowed in arrow expressions?

**Comment by rhdunn** [comment]
> > so `"fred" => upper-case()` could be now be written as `"fred" => upper-case`
> 
> Actually this would need to be:
> 
> **`"fred" => upper-case#1`**
> 
> Many years ago I asked if it would not be good to omit the arity in any case when a function definition has just a single overload. The answer is no, because if someone in the future adds another overload, then the expression without explicit arity becomes ambiguous

The arity here is not ambiguous, because the result after binding any parameters has to be a 1-arity function. Therefore, `upper-case` would always be the arity 1 variant. However, per my other reply, I think it is too much like an element path step, so is easy to be confusing when reading the expression.

**Comment by dnovatchev** [comment]
> > > so `"fred" => upper-case()` could be now be written as `"fred" => upper-case`
> > 
> > 
> > Actually this would need to be:
> > **`"fred" => upper-case#1`**
> > Many years ago I asked if it would not be good to omit the arity in any case when a function definition has just a single overload. The answer is no, because if someone in the future adds another overload, then the expression without explicit arity becomes ambiguous
> 
> The arity here is not ambiguous, because the result after binding any parameters has to be a 1-arity function. Therefore, `upper-case` would always be the arity 1 variant. However, per my other reply, I think it is too much like an element path step, so is easy to be confusing when reading the expression.

Agreed. Then this seems the only unambiguous expression (that is also simple and most readable):

  **`"fred" => upper-case#1`**

because even using the **"fn:"** or any other prefix cannot resolve the ambiguity, like in:

  **`"fred" => fn:upper-case`**
  
This may refer to an immediate child element of the context node, named **`fn:upper-case`** , and no one can  prohibit anybody from having an element named **`fn:upper-case`** .

Is my understanding correct, or am I missing something?

**Comment by ChristianGruen** [comment]
I agree that `upper-case` reads like an element name test.

I now believe the most natural representation would be the usage of the existing notation for inline functions without additional parentheses (if grammar permits)…

```xquery
$seq => function($n) { $n + 1 }()
```

…and, in a second step, a more compact syntax for online functions…

```xquery
fn($n) { $n + 1 }
{ $1 + 1 }
{ . + 1 }
...
```

…which can also be used with the (fat and, optional, thin) arrow operators.

@rhdunn If we allowed named function references, is my assumption correct that the result would be a function item, similar to when a partial function application is supplied…

```xquery
"fred" => substring(?)
```
…or would you expect the function item to be evaluated?

**Comment by rhdunn** [comment]
> @rhdunn If we allowed named function references, is my assumption correct that the result would be a function item, similar to when a partial function application is supplied…
> 
> ```xq
> "fred" => substring(?)
> ```
> 
> …or would you expect the function item to be evaluated?

I would consider `lower-case#1` to be equivalent to the current `"fred" => (lower-case(?))()` and `"fred" => (lower-case#1)()` case using parenthesized expressions. (That is, both bind to the 1-arity function.)

In your example, `?` is an argument in the `F` function call, so binds to the second argument, resolving `F` to a 2-arity function. I.e. your example is equivalent to `substring("fred", ?)`.

**Comment by ChristianGruen** [comment]
> I would consider `lower-case#1` to be equivalent to the current `"fred" => (lower-case(?))()` and `"fred" => (lower-case#1)()` case using parenthesized expressions.

Thanks. When thinking of the legal expression…

```
for $f in (lower-case#1, substring(?, 1))
return 'a' => $f()
```

…I believe it might come as a surprise if a RHS `lower-case#1` was invoked, whereas `substring(?, 1)` isn`t.

**Comment by rhdunn** [comment]
> > I would consider `lower-case#1` to be equivalent to the current `"fred" => (lower-case(?))()` and `"fred" => (lower-case#1)()` case using parenthesized expressions.
> 
> Thanks. When thinking of the legal expression…
> 
> ```
> for $f in (lower-case#1, substring(?, 1))
> return 'a' => $f()
> ```
> 
> …I believe it might come as a surprise if a RHS `lower-case#1` was invoked, whereas `substring(?, 1)` isn`t.

Right. In your example, I would expect both RHS expressions to be invoked with the supplied LHS argument.

**Comment by ChristianGruen** [comment]
With https://github.com/qt4cg/qtspecs/issues/435#issuecomment-1508228624 applied, there may be no need to further adjust the grammar.
